packages/contractkit/src/wrappers/Governance.ts

```diff
@@ -72,4 +136,298 @@ export class GovernanceWrapper extends BaseWrapper<Governance> {
+    const index = dequeue.findIndex((d) => d.isEqualTo(proposalID))
+    if (index === -1) {
+      throw new Error(`Proposal ${proposalID.toString()} not in dequeue`)
+    }
+    return index
+  }
+
+  // TODO: merge with SortedOracles/Election findLesserAndGreater
+  // proposalID is zero for revokes
+  private async findLesserAndGreaterAfterUpvote(proposalID: BigNumber, upvoter: Address) {
+    let queue = await this.getQueue()
+    let searchID = ZERO_BN
+
+    const upvoteRecord = await this.getUpvoteRecord(upvoter)
+    // does upvoter have a previous upvote?
+    if (upvoteRecord.id.isGreaterThan(ZERO_BN)) {
+      const proposalIdx = queue.findIndex((qp) => qp.id.isEqualTo(upvoteRecord.id))
+      // is previous upvote in queue?
+      if (proposalIdx !== -1) {
+        queue[proposalIdx].upvotes = queue[proposalIdx].upvotes.minus(upvoteRecord.weight)
+        searchID = upvoteRecord.id
+      }
+    }
+
+    // is upvoter targeting a valid proposal?
+    if (proposalID.isGreaterThan(ZERO_BN)) {
+      const proposalIdx = queue.findIndex((qp) => qp.id.isEqualTo(proposalID))
+      // is target proposal in queue?
+      if (proposalIdx !== -1) {
+        const weight = await this.getVoteWeight(upvoter)
+        queue[proposalIdx].upvotes = queue[proposalIdx].upvotes.plus(weight)
+        searchID = proposalID
+      } else {
+        throw new Error(`Proposal ${proposalID.toString()} not in queue`)
+      }
+    }
+
+    queue = queue.sort((a, b) => a.upvotes.comparedTo(b.upvotes))
+    const newIdx = queue.findIndex((qp) => qp.id.isEqualTo(searchID))
+
+    return {
+      lesserID: newIdx === 0 ? ZERO_BN : queue[newIdx - 1].id,
+      greaterID: newIdx === queue.length - 1 ? ZERO_BN : queue[newIdx + 1].id,
+    }
+  }
+
+  async upvote(proposalID: BigNumber, upvoter: Address) {
+    const exists = await this.proposalExists(proposalID)
+    if (!exists) {
+      throw new Error(`Proposal ${proposalID.toString()} does not exist`)
+    }
+    const { lesserID, greaterID } = await this.findLesserAndGreaterAfterUpvote(proposalID, upvoter)
+    return toTransactionObject(
+      this.kit,
+      this.contract.methods.upvote(
+        proposalID.toString(),
+        lesserID.toString(),
+        greaterID.toString()
+      )
+    )
+  }
+
+  async revokeUpvote(upvoter: Address) {
+    const { id } = await this.getUpvoteRecord(upvoter)
+    if (!id.isGreaterThan(ZERO_BN)) {
+      throw new Error(`Voter ${upvoter} has no upvote to revoke`)
+    }
+    const { lesserID, greaterID } = await this.findLesserAndGreaterAfterUpvote(ZERO_BN, upvoter)
+    return toTransactionObject(
+      this.kit,
+      this.contract.methods.revokeUpvote(lesserID.toString(), greaterID.toString())
+    )
+  }
+
+  async approve(proposalID: BigNumber) {
+    const proposalIndex = await this.getDequeueIndex(proposalID)
+    return toTransactionObject(
+      this.kit,
+      this.contract.methods.approve(proposalID.toString(), proposalIndex)
```

Comment:
we should use `toFixed()` for BigNumbers. Sometimes, toString() returns `1e10` format